File: Memba.WebControls/JavaScript/ProgressReport.js
/*
 * Copyright © 2005-2008 Memba SA. All rights reserved.
 * 
 * This file is part of Velodoc XP Edition.
 * 
 * Velodoc XP Edition is open-source software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 2 of the License, or (at your option) any later version.
 * 
 * Velodoc XP Edition is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along with Velodoc XP Edition.
 * If not, see <http://www.gnu.org/licenses/>.
 * 
 * For more information, please contact Memba at <http://www.memba.com>.
 * You can find more information about Velodoc at <http://www.velodoc.com> and <http://www.velodoc.com/help>.
 * You can try Velodoc online at <http://www.velodoc.net>.
 *
*/ 
Type.registerNamespace('Memba.WebControls');

//--------------------------------------------
//ProgressReport Control
//--------------------------------------------
Memba.WebControls.ProgressReport = function(element)
{
    Sys.Debug.trace("ProgressReport Constructor");
    
    this._DefaulText = null;
    this._TextFormat = null
    this._HandlerUrl = null;
    this._Interval = null;
    this._Timeout = null;
    this._TimeoutID = null;
    
    this._CssClass = null;
    this._BarCssClass = null;
    this._FillerCssClass = null;
    this._BarCssClass = null;
    
    this._MainDiv = null;
    this._BarDiv = null;
    this._FillerDiv = null;
    this._TextDiv = null;
    
    this._CurrentRequest = null;
    this._IsComplete = null;
    this._IsStarted = false;
           
    Memba.WebControls.ProgressReport.initializeBase(this, [element]);
}

Memba.WebControls.ProgressReport.prototype =
{
    get_defaultText : function()
    {
        return this._DefaultText;
    },

    set_defaultText : function(value)
    {
        if (this._DefaultText != value)
        {
            this._DefaultText = value;
            if (this.get_isInitialized() && (this._TextDiv))
                this._TextDiv.innerHTML = value;

            // Raise the propertyChanged event
            this.raisePropertyChanged('defaultText');
        }
    },
    
    get_textFormat : function()
    {
        return this._TextFormat;
    },

    set_textFormat : function(value)
    {
        if (this._TextFormat != value)
        {
            this._TextFormat = value;

            // Raise the propertyChanged event
            this.raisePropertyChanged('textFormat');
        }
    },
    
    get_handlerUrl : function()
    {
        return this._HandlerUrl;
    },
 
    set_handlerUrl : function(value)
    {
        if (this._HandlerUrl != value)
        {
            this._HandlerUrl = value;

            // Raise the propertyChanged event
            this.raisePropertyChanged('handlerUrl');
        }
    },

    get_interval : function()
    {
        return this._Interval;
    },
 
    set_interval : function(value)
    {
        if (this._Interval != value)
        {
            this._Interval = value;

            // Raise the propertyChanged event
            this.raisePropertyChanged('interval');
        }
    },

    get_timeout : function()
    {
        return this._Timeout;
    },
 
    set_timeout : function(value)
    {
        if (this._Timeout != value)
        {
            this._Timeout = value;

            // Raise the propertyChanged event
            this.raisePropertyChanged('timeout');
        }
    },

    get_cssClass : function()
    {
        return this._CssClass;
    },

    set_cssClass : function(value)
    {
        if (this._CssClass != value)
        {
            this._CssClass = value;
            if (this.get_isInitialized() && (this._MainDiv))
                this._MainDiv.className = value;

            // Raise the propertyChanged event
            this.raisePropertyChanged('cssClass');
        }
    },
    
    get_barCssClass : function()
    {
        return this._BarCssClass;
    },

    set_barCssClass : function(value)
    {
        if (this._BarCssClass != value)
        {
            this._BarCssClass = value;
            if (this.get_isInitialized() && (this._BarDiv))
                this._BarDiv.className = value;

            // Raise the propertyChanged event
            this.raisePropertyChanged('barCssClass');
        }
    },
    
    get_fillerCssClass : function()
    {
        return this._FillerCssClass;
    },

    set_fillerCssClass : function(value)
    {
        if (this._FillerCssClass != value)
        {
            this._FillerCssClass = value;
            if (this.get_isInitialized() && (this._FillerDiv))
                this._FillerDiv.className = value;

            // Raise the propertyChanged event
            this.raisePropertyChanged('fillerCssClass');
        }
    },
    
    get_textCssClass : function()
    {
        return this._TextCssClass;
    },

    set_textCssClass : function(value)
    {
        if (this._TextCssClass != value)
        {
            this._TextCssClass = value;
            if (this.get_isInitialized() && (this._Label))
                this._TextDiv.className = value;

            // Raise the propertyChanged event
            this.raisePropertyChanged('textCssClass');
        }
    },
    
    // error event
    add_error: function(handler) {
        /// <summary>Adds a event handler for the error event.</summary>
        /// <param name="handler" type="Function">The handler to add to the event.</param>
        this.get_events().addHandler("error", handler);
    },
    remove_error: function(handler) {
        /// <summary>Removes a event handler for the error event.</summary>
        /// <param name="handler" type="Function">The handler to remove from the event.</param>
        this.get_events().removeHandler("error", handler);
    },

    // complete event
    add_complete: function(handler) {
        /// <summary>Adds a event handler for the complete event.</summary>
        /// <param name="handler" type="Function">The handler to add to the event.</param>
        this.get_events().addHandler("complete", handler);
    },
    remove_complete: function(handler) {
        /// <summary>Removes a event handler for the complete event.</summary>
        /// <param name="handler" type="Function">The handler to remove from the event.</param>
        this.get_events().removeHandler("complete", handler);
    },
    
    start : function()
    {
        Sys.Debug.trace("Starting progress report");
        this._IsStarted = true;
        this._updateProgress();
    },
    
    stop : function()
    {
        Sys.Debug.trace("Stopping progress report");
        this._IsStarted = false;       
        //Abort request in progress 
        if((typeof(this._CurrentRequest) != "undefined") && (this._CurrentRequest != null)
            && (typeof(this._CurrentRequest.get_executor()) != "undefined") && (this._CurrentRequest.get_executor() != null))
            this._CurrentRequest.get_executor().abort();
    },
    
    //XmlHttp call to progress handler
    _updateProgress : function()
    {
        Sys.Debug.trace("Ajax call to update progress status");

        //No progress handler
        if ((typeof(this._HandlerUrl) == "undefined") || (this._HandlerUrl == null) || (this._HandlerUrl == ''))
            return;
        
        //Progress not visible
        if ((this._FillerDiv == null) && (this._TextDiv == null))
            return;
        
        //Progress not started, you need to call start(), not _updateProgress()
        if (!this._IsStarted)
            return;
        
        //Progress is complete
        if (this._IsComplete)
            return;
    
        //Abort request in progress to avoid queuing many identical requests 
        if((typeof(this._CurrentRequest) != "undefined") && (this._CurrentRequest != null)
            && (typeof(this._CurrentRequest.get_executor()) != "undefined") && (this._CurrentRequest.get_executor() != null))
            this._CurrentRequest.get_executor().abort();
        
        Sys.Debug.trace("Calling " + this._HandlerUrl);
        
        //Execute new request
        this._CurrentRequest = new Sys.Net.WebRequest();
        this._CurrentRequest.set_url(this._HandlerUrl);
        this._CurrentRequest.set_httpVerb("GET");
        this._CurrentRequest.set_timeout(this._Timeout);
        this._CurrentRequest.add_completed(Function.createDelegate(this, this._updateProgressCallback));
        this._CurrentRequest.invoke();
    
        //Call again later. We plan the next call here in case any callback does not execute 
        Sys.Debug.trace("Planning next call in " + this._Interval + "ms");
        this._TimeoutID = window.setTimeout(Function.createDelegate(this, this._updateProgress) , this._Interval);
    },
    
    //Asynchronous callback
    _updateProgressCallback : function(executor, eventArgs) 
    {    
        Sys.Debug.trace("Asynchronous callback to update progress status");
        
        //Progress not started
        if (!this._IsStarted)
            return;
        
        if(executor.get_responseAvailable()) 
        {
            Sys.Debug.trace("The callback did respond");
            
            try
            {
                var progressData = Sys.Serialization.JavaScriptSerializer.deserialize(executor.get_responseData());
                if (progressData)
                {
                    if((progressData.ErrorMessage != null) && (progressData.ErrorMessage != ''))
                    {
                        Sys.Debug.trace("There is an error: " + progressData.ErrorMessage);
                        
                        this._IsComplete = true;
                        //Raise the error event
                        var eventHandler = this.get_events().getHandler("error");
                        if (eventHandler)
                            eventHandler(this, progressData.ErrorMessage);
                    }
                    else
                    {
                        Sys.Debug.trace("We are progressing: " + progressData.FillRatio);

                        if (progressData.FillRatio)
                        {
                            var width = progressData.FillRatio;
                            
                            if(this._FillerDiv)
                            {
                                
                                //this._FillerDiv.style.width = width;
                                
                                var _updateFillerDiv = Function.createDelegate(this, function() { 
                                    this._FillerDiv.style.width = width;
                                    Sys.Debug.trace("Filler width is now " + this._FillerDiv.style.width);
                                });

                                window.setTimeout(_updateFillerDiv, 0);
                            }
                        }
                        
                        if((progressData.TextValues) && (progressData.TextValues instanceof Array))
                        {
                            if (progressData.TextValues.length > 0)
                            {
                                if (this._TextDiv)
                                {
                                    //using String.localeFormat does not work, see _toFormattedString below
                                    //var text = String.localeFormat(this._TextFormat, progressData.TextValues);
                                    var text = this._toFormattedString(this._TextFormat, progressData.TextValues);
                                    
                                    //It is probably simpler to do the following but
                                    //1) On one hand, some articles say JavaScript browser implementations are single threaded
                                    //2) On the other hand, other articles argue that setTimeout provides a "multithreaded effect"
                                    //TODO: Probably needs to be reviewed as we get more information.
                                    //this._TextDiv.innerHTML = eval(text);

                                    var _updateTextDiv = Function.createDelegate(this, function() { 
                                        Sys.Debug.trace("Updating text with " + text);
                                        this._TextDiv.innerHTML = text; //eval(text);
                                    });
                                
                                    window.setTimeout(_updateTextDiv, 0);
                                }
                            }
                        }
                        
                        //If Upload is complete
                        if (progressData.IsComplete)
                        {
                            Sys.Debug.trace("We are complete");

                            this._IsComplete = true;

                            this.stop();

                            //Raise the complete event
                            var eventHandler = this.get_events().getHandler("complete");
                            if (eventHandler)
                                eventHandler(this, progressData.FillRatio);
                        }
                    }
                }
            }
            catch(e)
            {
            }
        }
    },

    //This function is derived from String._toFormattedString which I could not get to work as expected when args is an array
    _toFormattedString : function(format, args) {
        /// <param name="format" type="String"></param>
        /// <param name="args" parameterArray="true" mayBeNull="true"></param>
        /// <returns type="String"></returns>
        var e = Function._validateParams(arguments, [
            {name: "format", type: String},
            {name: "args", mayBeNull: true, parameterArray: true}
        ]);
        if (e) throw e;

        //alert(arguments.length) displays 2
        //This is the problem of the String.localFormat() function which works on the basis
        //that arguments[0] = format and arguments[i+1] = args[i]
        //when in fact we have arguments[0] = format and arguments[1] = args.

        var useLocale = true;
        var result = '';

        for (var pos=0;;) {
        
            // Find the next opening or closing brace
            var open = format.indexOf('{', pos);
            var close = format.indexOf('}', pos);
            if ((open < 0) && (close < 0)) {
                // Not found: copy the end of the string and break
                result += format.slice(pos);
                break;
            }
            if ((close > 0) && ((close < open) || (open < 0))) {
                // Closing brace before opening is an error
                if (format.charAt(close + 1) !== '}') {
                    throw Error.argument('format', Sys.Res.stringFormatBraceMismatch);
                }
                result += format.slice(pos, close + 1);
                pos = close + 2;
                continue;
            }

            // Copy the string before the brace
            result += format.slice(pos, open);
            pos = open + 1;

            // Check for double braces (which display as one and are not arguments)
            if (format.charAt(pos) === '{') {
                result += '{';
                pos++;
                continue;
            }

            // at this point we have a valid opening brace, which should be matched by a closing brace.
            if (close < 0) throw Error.argument('format', Sys.Res.stringFormatBraceMismatch);

            // Find the closing brace

            // Get the string between the braces, and split it around the ':' (if any)
            var brace = format.substring(pos, close);
            var colonIndex = brace.indexOf(':');
            var argNumber = parseInt((colonIndex < 0)? brace : brace.substring(0, colonIndex)); // + 1;
            if (isNaN(argNumber)) throw Error.argument('format', Sys.Res.stringFormatInvalid);
            var argFormat = (colonIndex < 0)? '' : brace.substring(colonIndex + 1);

            //alert(argNumber.toString() + " -> " + args[argNumber]);
            var arg = args[argNumber];
            if (typeof(arg) === "undefined" || arg === null) {
                arg = '';
            }

            // If it has a toFormattedString method, call it.  Otherwise, call toString()
            if (arg.toFormattedString) {
                result += arg.toFormattedString(argFormat);
            }
            else if (useLocale && arg.localeFormat) {
                result += arg.localeFormat(argFormat);
            }
            else if (arg.format) {
                result += arg.format(argFormat);
            }
            else
                result += arg.toString();

            pos = close + 1;
        }

        return result;
    },

    initialize : function()
    {
        Sys.Debug.trace("ProgressReport initialization");
        
        Memba.WebControls.ProgressReport.callBaseMethod(this, 'initialize');
        
        this._MainDiv = this.get_element();
        if(this._MainDiv)
            this._MainDiv.className = this._CssClass;
        this._BarDiv = $get(this.get_id() + "_Bar");
        if(this._BarDiv)
            this._BarDiv.className = this._BarCssClass;
        this._FillerDiv = $get(this.get_id() + "_Filler");
        if (this._FillerDiv)
            this._FillerDiv.className = this._FillerCssClass;
        this._TextDiv = $get(this.get_id() + "_Text");
        if (this._TextDiv)
        {
            this._TextDiv.className = this._TextCssClass;
            this._TextDiv.innerHTML = this._DefaultText;
        }
    },
    
    dispose : function()
    {
        Sys.Debug.trace("ProgressReport disposal");
        
        Memba.WebControls.ProgressReport.callBaseMethod(this, 'dispose');
    }
}

Memba.WebControls.ProgressReport.registerClass('Memba.WebControls.ProgressReport', Sys.UI.Control);

if (typeof (Sys) != 'undefined')
{
    Sys.Application.notifyScriptLoaded();
}

File: Memba.WebControls/ProgressReport.cs
/*
 * Copyright © 2005-2008 Memba SA. All rights reserved.
 * 
 * This file is part of Velodoc XP Edition.
 * 
 * Velodoc XP Edition is open-source software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 2 of the License, or (at your option) any later version.
 * 
 * Velodoc XP Edition is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along with Velodoc XP Edition.
 * If not, see <http://www.gnu.org/licenses/>.
 * 
 * For more information, please contact Memba at <http://www.memba.com>.
 * You can find more information about Velodoc at <http://www.velodoc.com> and <http://www.velodoc.com/help>.
 * You can try Velodoc online at <http://www.velodoc.net>.
 *
*/
using System;
using System.Collections.Generic;
using System.Text;
using System.Web; //AspNetHostingPermission, AspNetHostingPermissionLevel
using System.Web.UI; //Control, IScriptControl, WebResource, Themeable
using System.Web.UI.Design; //UrlEditor
using System.Web.UI.WebControls;
using System.Drawing.Design; //UITypeEditor
using System.ComponentModel; //Category, Description, DefaultProperty
using System.Security.Permissions; //SecurityAction


//See: http://www.asp.net/AJAX/Documentation/Live/tutorials/IScriptControlTutorial1.aspx

[assembly: WebResource("Memba.WebControls.JavaScript.ProgressReport.js", "text/javascript")]

namespace Memba.WebControls
{
    [AspNetHostingPermission(SecurityAction.LinkDemand, Level = AspNetHostingPermissionLevel.Minimal)]
    [AspNetHostingPermission(SecurityAction.InheritanceDemand, Level = AspNetHostingPermissionLevel.Minimal)]
    [DefaultProperty("DefaultText")]
    [Themeable(true)]
    //[Designer("...")]
    //[ParseChildren(true)]
    //[PersistChildren(false)]
    public class ProgressReport : Control, IScriptControl
    {
        private const int DEFAULT_TIMEOUT = 3000;
        private const int DEFAULT_INTERVAL = 3000;

        #region Private Members
        private string _DefaultText;
        private string _TextFormat;
        private string _HandlerUrl;
        private int _Interval;
        private int _Timeout;
        private string _CssClass;
        private string _BarCssClass;
        private string _FillerCssClass;
        private string _TextCssClass;
        private Unit _Width;
        private Unit _Height;
        //private Unit _BarHeight;
        #endregion

        #region Constructor
        /// <summary>
        /// Constructor
        /// </summary>
        public ProgressReport()
            : base()
        {
            _Interval = DEFAULT_INTERVAL;
            _Timeout = DEFAULT_TIMEOUT;
            //_BarHeight = new Unit("10px");
        }
        #endregion

        #region Property Accessors
        /// <summary>
        /// The text displayed before the handler is called
        /// </summary>
        [Category("Appearance")]
        [Description("The text displayed before the handler is called")]
        [Themeable(false)]
        [Localizable(true)]
        [DefaultValue("")]
        public string DefaultText
        {
            get { return _DefaultText; }
            set { _DefaultText = value; }
        }
        /// <summary>
        /// A composite format string displayed below the progress bar
        /// </summary>
        [Category("Appearance")]
        [Description("A composite format string displayed below the progress bar")]
        [Themeable(false)]
        [Localizable(true)]
        [DefaultValue("")]
        public string TextFormat
        {
            get { return _TextFormat; }
            set { _TextFormat = value; }
        }
        /// <summary>
        /// Url of the handler returning upload progress status
        /// </summary>
        [Category("Behavior")]
        [Description("Url of the handler returning upload progress status")]
        [DefaultValue("")]
        [Editor(typeof(UrlEditor), typeof(UITypeEditor))]
        public string HandlerUrl
        {
            get { return _HandlerUrl; }
            set { _HandlerUrl = value.Trim(); }
        }
        /// <summary>
        /// Interval at which recurrently calling the handler
        /// </summary>
        [Category("Behavior")]
        [Description("Interval at which recurrently calling the handler")]
        [DefaultValue(ProgressReport.DEFAULT_INTERVAL)]
        public int Interval
        {
            get { return _Interval; }
            set { _Interval = value; }
        }
        /// <summary>
        /// Interval at which recurrently calling the handler
        /// </summary>
        [Category("Behavior")]
        [Description("Timeout of the handler call")]
        [DefaultValue(ProgressReport.DEFAULT_TIMEOUT)]
        public int Timeout
        {
            get { return _Timeout; }
            set { _Timeout = value; }
        }
        /// <summary>
        /// CSS class applied to the container div
        /// </summary>
        [Category("Appearance")]
        [Description("CSS class applied to the container div")]
        [DefaultValue("")]
        public string CssClass
        {
            get { return _CssClass; }
            set { _CssClass = value.Trim(); }
        }
        /// <summary>
        /// CSS class applied to the outer div of the progress bar
        /// </summary>
        [Category("Appearance")]
        [Description("CSS class applied to the outer div of the progress bar")]
        [DefaultValue("")]
        public string BarCssClass
        {
            get { return _BarCssClass; }
            set { _BarCssClass = value.Trim(); }
        }
        /// <summary>
        /// CSS class applied to the inner div of the progress bar
        /// </summary>
        [Category("Appearance")]
        [Description("CSS class applied to the inner div of the progress bar")]
        [DefaultValue("")]
        public string FillerCssClass
        {
            get { return _FillerCssClass; }
            set { _FillerCssClass = value.Trim(); }
        }
        /// <summary>
        /// CSS class applied to the text displayed below the progress bar
        /// </summary>
        [Category("Appearance")]
        [Description("CSS class applied to the text displayed below the progress bar")]
        [DefaultValue("")]
        public string TextCssClass
        {
            get { return _TextCssClass; }
            set { _TextCssClass = value.Trim(); }
        }
        /// <summary>
        /// Width of the upload progress control
        /// </summary>
        [Category("Layout")]
        [Description("Width of the upload progress control")]
        [DefaultValue(typeof(Unit), "")]
        public Unit Width
        {
            get { return _Width; }
            set { _Width = value; }
        }
        /// <summary>
        /// Height of the upload progress control
        /// </summary>
        [Category("Layout")]
        [Description("Height of the upload progress control")]
        [DefaultValue(typeof(Unit), "")]
        public Unit Height
        {
            get { return _Height; }
            set { _Height = value; }
        }
        /// <summary>
        /// Height of the progress bar
        /// </summary>
        //[Category("Layout")]
        //[Description("Height of the progress bar")]
        //[DefaultValue(typeof(Unit), "10px")]
        //public Unit BarHeight
        //{
        //    get { return _BarHeight; }
        //    set { _BarHeight = value; }
        //}
        /// <summary>
        /// 
        /// </summary>
        [Browsable(true)]
        public override bool EnableTheming
        {
            get { return base.EnableTheming; }
            set { base.EnableTheming = value; }
        }
        /// <summary>
        /// 
        /// </summary>
        [Browsable(true)]
        public override string SkinID
        {
            get { return base.SkinID; }
            set { base.SkinID = value; }
        }
        #endregion

        #region Control Events
        /// <summary>
        /// 
        /// </summary>
        /// <param name="e"></param>
        protected override void OnPreRender(EventArgs e)
        {
            if (!this.DesignMode)
            {
                // Test for ScriptManager and register if it exists
                ScriptManager objScriptManager = ScriptManager.GetCurrent(Page);

                if (objScriptManager == null)
                    throw new HttpException("A ScriptManager control must exist on the current page.");

                //Register control with ScriptManager
                objScriptManager.RegisterScriptControl(this);
            }

            base.OnPreRender(e); //was on top of the method

        }
        /// <summary>
        /// 
        /// </summary>
        /// <param name="writer"></param>
        protected override void Render(HtmlTextWriter writer)
        {
            /*
            <div id="X"> 
                <div id="X_Bar">
                    <div id="X_Filler" style="width:0%"></div>
                </div>
                <div id="X_Text">bla bla</div>  
            </div> 
            */

            if (this.DesignMode || this.Visible)
            {
                writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID);
                if (!String.IsNullOrEmpty(_CssClass))
                    writer.AddAttribute(HtmlTextWriterAttribute.Class, _CssClass);
                if (!_Width.IsEmpty)
                    writer.AddStyleAttribute(HtmlTextWriterStyle.Width, _Width.ToString());
                if (!_Height.IsEmpty)
                    writer.AddStyleAttribute(HtmlTextWriterStyle.Height, _Height.ToString());
                writer.RenderBeginTag(HtmlTextWriterTag.Div);

                //Progress bar
                writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID + "_Bar");
                if (!String.IsNullOrEmpty(_BarCssClass))
                    writer.AddAttribute(HtmlTextWriterAttribute.Class, _BarCssClass);
                writer.RenderBeginTag(HtmlTextWriterTag.Div);
                writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID + "_Filler");
                writer.AddStyleAttribute(HtmlTextWriterStyle.Width, "0%");
                if (!String.IsNullOrEmpty(_FillerCssClass))
                    writer.AddAttribute(HtmlTextWriterAttribute.Class, _FillerCssClass);
                writer.RenderBeginTag(HtmlTextWriterTag.Div);
                writer.RenderEndTag(); //div
                writer.RenderEndTag(); //div

                //Text
                writer.AddAttribute(HtmlTextWriterAttribute.Id, this.ClientID + "_Text");
                if (!String.IsNullOrEmpty(_TextCssClass))
                    writer.AddAttribute(HtmlTextWriterAttribute.Class, _TextCssClass);
                writer.RenderBeginTag(HtmlTextWriterTag.Div);
                writer.Write(_DefaultText);
                writer.RenderEndTag(); //div

                writer.RenderEndTag(); //div
            }

            //Register the Ajax extension properties of the control
            if (!this.DesignMode)
            {
                ScriptManager.GetCurrent(this.Page).RegisterScriptDescriptors(this);
            }

            base.Render(writer); //Was on top of method
        }
        #endregion

        #region IScriptControl interface implementation
        /// <summary>
        /// Protected GetScriptReferences which should be overriden n derived classes
        /// </summary>
        /// <returns></returns>
        protected virtual IEnumerable<ScriptReference> GetScriptReferences()
        {
            yield return new ScriptReference(
                this.Page.ClientScript.GetWebResourceUrl(
                    this.GetType(),
                    "Memba.WebControls.JavaScript.ProgressReport.js"));
        }
        /// <summary>
        /// Protected GetScriptDescriptors which should be overriden n derived classes
        /// </summary>
        /// <returns></returns>
        protected virtual IEnumerable<ScriptDescriptor> GetScriptDescriptors()
        {
            ScriptControlDescriptor objScriptControlDescriptor = new ScriptControlDescriptor("Memba.WebControls.ProgressReport", ClientID);
            objScriptControlDescriptor.AddProperty("defaultText", this.DefaultText);
            objScriptControlDescriptor.AddProperty("textFormat", this.TextFormat);
            objScriptControlDescriptor.AddProperty("handlerUrl", (String.IsNullOrEmpty(this.HandlerUrl) ? null : this.ResolveClientUrl(this.HandlerUrl)));
            objScriptControlDescriptor.AddProperty("interval", this.Interval);
            objScriptControlDescriptor.AddProperty("timeout", this.Timeout);
            objScriptControlDescriptor.AddProperty("cssClass", this.CssClass);
            objScriptControlDescriptor.AddProperty("barCssClass", this.BarCssClass);
            objScriptControlDescriptor.AddProperty("fillerCssClass", this.FillerCssClass);
            objScriptControlDescriptor.AddProperty("textCssClass", this.TextCssClass);

            yield return objScriptControlDescriptor;
        }
        /// <summary>
        /// Public interface for GetScriptReferences
        /// </summary>
        /// <returns></returns>
        IEnumerable<ScriptReference> IScriptControl.GetScriptReferences()
        {
            return GetScriptReferences();
        }
        /// <summary>
        /// Public interface for GetScriptDescriptors
        /// </summary>
        /// <returns></returns>
        IEnumerable<ScriptDescriptor> IScriptControl.GetScriptDescriptors()
        {
            return GetScriptDescriptors();
        }
        #endregion
    }
}
